### PR TITLE
Create AssignedCoursesAndScripts concern

### PR DESCRIPTION
**Pull request by nicklathe**
First of all, this PR is bigger than anticipated. Initially I wanted this to be two concerns, AssignedCourses and AssignedScripts, however there is too much overlap in the methods to split it up cleanly.

This PR adds a new AssignedCoursesAndScripts concern, which is included in the User model. It moves a number of course, script, and unit related methods to the new concern, in an ongoing effort to modularize the User model by use of concerns.

Some of the methods were un-tested, and in writing unit tests for them it was discovered that a few of the methods either didn't handle edge cases well, or flat out didn't behave as expected. The details are covered in comments in the PR.

Finally, this PR also adds a number of unit tests for previously untested methods.

## Links

This concern combines two Jira tickets:
- https://codedotorg.atlassian.net/browse/P20-1473
- https://codedotorg.atlassian.net/browse/P20-1474

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/models/concerns/user/assigned_courses_and_scripts.rb [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+  end
+
+  # Checks if there are any launched scripts or courses assigned to the user.
+  # @return [Boolean]
+  def assigned_course_or_script?
+    assigned_courses.any? || any_visible_assigned_scripts?
+  end
+
+  # Query to get the user_script the user was most recently assigned.
+  def most_recently_assigned_user_script
```
Comment (nicklathe) [inline dashboard/app/models/concerns/user/assigned_courses_and_scripts.rb:49]: This method was untested. When I wrote a unit test for it, I realized it was not returning the most recently assigned script.

Old method
```
  # Query to get the user_script the user was most recently assigned.
  def most_recently_assigned_user_script
    user_scripts.
      where("assigned_at").
      order(assigned_at: :desc).
      first
  end
```

What's happening is that in [user.rb](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/models/user.rb#L290), we have:
`has_many :user_scripts, -> {order Arel.sql("-completed_at asc, greatest(coalesce(started_at, 0), coalesce(assigned_at, 0), coalesce(last_progress_at, 0)) desc, user_scripts.id asc")}`.

This means that anytime you call `user_scripts`, it automatically includes that complex ordering, even if you call `.order(...)` afterward.

The fix is to use `.reorder` here, which ensures all previous ORDER BY clauses are removed, including the one defined in the association.

Additionally, `.where(assigned_at)` was refactored to `.where.not(assigned_at: nil)` to more explicitly state we are looking for records that have actually been assigned.

File: dashboard/app/models/concerns/user/assigned_courses_and_scripts.rb [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    # UserScript doesn't record the UnitGroup the user was in, so we will assume
+    # it is the most recently created section.
+    section = sections_as_student.select {|s| !s.hidden && s.script_id == unit.id}.last
+    Queries::Courses.unit_group_unit(unit, section&.unit_group)
+  end
+
+  # Get script object of the user_script the user was most recently
+  # assigned.
+  def most_recently_assigned_script
+    most_recently_assigned_user_script&.script
```
Comment (nicklathe) [inline dashboard/app/models/concerns/user/assigned_courses_and_scripts.rb:69]: Added the Safe Navigation Operator (`&`) here, which was missing in the old method.

File: dashboard/app/models/concerns/user/assigned_courses_and_scripts.rb [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+  # Get script object of the user_script the user made the most recent
+  # progress in.
+  def script_with_most_recent_progress
+    user_script_with_most_recent_progress&.script
+  end
+
+  # Check if the user's most recently-assigned script is the same one
+  # that they've most recently made progress in.
+  def most_recent_progress_in_recently_assigned_script?
```
Comment (nicklathe) [inline dashboard/app/models/concerns/user/assigned_courses_and_scripts.rb:118]: This was another (like many) un-tested method that didn't behave as expected once I started writing tests.

The old method (as detailed below) would incorrectly return true if both the called methods returned nil, because `nil == nil # true`
```
  def most_recent_progress_in_recently_assigned_script?
    script_with_most_recent_progress == most_recently_assigned_script
  end
```

This refactored method now will return false if either of those methods return nil, before comparing their returned values.

File: dashboard/app/models/concerns/user/assigned_courses_and_scripts.rb [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        all_scripts.concat(section.unit_group.default_units)
+      end
+    end
+
+    all_scripts
+  end
+
+  # Query to get the user_script the user made the most recent progress
+  # in.
+  def user_script_with_most_recent_progress
```
Comment (nicklathe) [inline dashboard/app/models/concerns/user/assigned_courses_and_scripts.rb:103]: This method was also refactored to account for the ordering issue outlined above.

File: dashboard/app/models/concerns/user/assigned_courses_and_scripts.rb [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+    all_scripts = []
+    all_sections.each do |section|
+      if section.script.present?
+        all_scripts << section.script
+      elsif section.unit_group.present?
+        all_scripts.concat(section.unit_group.default_units)
+      end
+    end
+
+    all_scripts
```
[the remark this answers is not in the document]
Comment (nicklathe) [reply dashboard/app/models/concerns/user/assigned_courses_and_scripts.rb:98]: Maybe. It probably doesn't hurt to filter for uniqueness here, but I'm not sure if we would end up with dup scripts in the real world?

File: dashboard/app/models/concerns/user/assigned_courses_and_scripts.rb [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+  # progress in.
+  def script_with_most_recent_progress
+    user_script_with_most_recent_progress&.script
+  end
+
+  # Check if the user's most recently-assigned script is the same one
+  # that they've most recently made progress in.
+  def most_recent_progress_in_recently_assigned_script?
+    return false if script_with_most_recent_progress.nil? || most_recently_assigned_script.nil?
+    script_with_most_recent_progress == most_recently_assigned_script
```
Comment (carl-codeorg) [inline dashboard/app/models/concerns/user/assigned_courses_and_scripts.rb:120]: Is it safer to compare IDs? I'm always wary of using equality operators on whole objects.

**Review by carl-codeorg — APPROVED** [review]
This seems like a pretty messy area of the model, so thanks for going in and cleaning it up a bit. I'm not an expert on courses/scripts, but from my perspective this looks good. The only comment I had is that these functions do a lot of comparisons of ActiveRecord objects when I normally consider it safer to compare IDs. It's possible that it just works to compare entire objects in Rails though.

File: dashboard/app/models/concerns/user/assigned_courses_and_scripts.rb (hunk 6 above)
Comment (nicklathe) [reply dashboard/app/models/concerns/user/assigned_courses_and_scripts.rb:120]: That's a great question. It seems safe to compare resources in Active Record. I found some stack overflow things, but this seems to be the most "source of truth" I can find. I'm curious what is deemed best practice in our codebase. https://apidock.com/rails/v3.2.13/ActiveResource/Base/%3D%3D
